openedx-unsupported/edx-analytics-pipeline
# SE-1032 Adds option to store anonymous user events in module_engagement hive table

Since the Public Courseware feature was released, our client's engagement data has dropped significantly, and so they would like a way to count the anonymous hits to this content.  They use the per-learner module_engagement hive tables for their own internal reporting, and so only need the data recorded there.

Add this configuration to store anonymous user records under the given replacement username.  Best to choose one that is an invalid username, to avoid accidental overlap with real users, e.g.
```cfg
[module-engagement]
store_anonymous_username = ANONYMOUS USER
```

**JIRA tickets**:  [OSPR-3659](https://openedx.atlassian.net/browse/OSPR-3659)

**Merge deadline**: None, though we would appreciate feedback on the approach ASAP so we can decide whether to merge to our client's branch now.

**Testing instructions**:

Set up:
1. Set up the [Analytics Pipeline docker devstack](https://github.com/edx/devstack#getting-started-on-analytics)
1. Check out this branch and install it: 
   ```bash
   make analytics-pipeline-shell
   # From devstack/Makefile#L300
   sudo chown -R hadoop:hadoop /edx/app/analytics_pipeline
   source /edx/app/hadoop/.bashrc
   make install
   ```

Automated testing:
1. Run this in the analytics-pipeline shell:
   ```bash
   make develop-local
   make docker-test-acceptance-local  ONLY_TESTS=edx.analytics.tasks.tests.acceptance.test_module_engagement
   ```

Manual testing:

1. Copy the anonymous user tracking logs into the hdfs /data dir:
   * [anon_tracking.log](https://github.com/edx/edx-analytics-pipeline/files/3289758/anon_tracking.log)
   * [honor_tracking.log](https://github.com/edx/edx-analytics-pipeline/files/3289759/honor_tracking.log)
   ```
   hdfs dfs -put anon_tracking.log /data/
   hdfs dfs -put honor_tracking.log /data/
   hdfs dfs -ls /data
   Found 2 items
   -rw-r--r--   3 hadoop hadoop      12897 2019-06-11 06:32 /data/anon_tracking.log
   -rw-r--r--   3 hadoop hadoop      12972 2019-06-11 06:42 /data/honor_tracking.log
   ```
1. Check the hive database  -- a clean install won't have any tables in it, but even if you do, there shouldn't be any anonymous user records in the `module_engagement` table yet.
   ```bash
   hive
   hive> show tables;
   hive> select * from module_engagement;
   ```
1. Update `/edx/app/analytics_pipeline/analytics_pipeline/override.cfg` for the module-engagement tasks, ignoring anonymous user records by default:
   ```cfg
   [module-engagement]
   allow_empty_insert = true
   ```
1. Process the tracking data loaded into hdfs above.
   ```bash
   START_DATE=2019-06-10; END_DATE=2019-06-12;
   # Load enrollments for the interval
   launch-task ImportEnrollmentsIntoMysql --interval $START_DATE-$END_DATE --local-scheduler
   # Process per-learner data for the interval
   launch-task ModuleEngagementWorkflowTask --date $END_DATE --local-scheduler
   ``` 
1. Check that only the honor user records are in hive
   ```bash
   hive> select * from module_engagement;
   edX/DemoX/Demo_Course_2 honor   2019-06-10      video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f    viewed 2       2019-06-10
   edX/DemoX/Demo_Course   honor   2019-06-10      video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f    viewed 2       2019-06-10
   Time taken: 1.929 seconds, Fetched: 4 row(s)
   ``` 
1. Update `override.cfg` to store anonymous users under the username, `ANONYMOUS USER`
   ```cfg
   [module-engagement]
   store_anonymous_username = ANONYMOUS USER
   allow_empty_insert = true
   ```
1. Overwrite the module engagement data to include these anonymous user records
   ```bash
   launch-task ModuleEngagementIntervalTask --interval $START_DATE-$END_DATE --local-scheduler --overwrite-from-date $START_DATE  --overwrite-mysql
   ```
1. Check that there are now both honor and anonymous user records present in the table.
   ```bash
   hive> select * from module_engagement;
   edX/DemoX/Demo_Course_2 ANONYMOUS USER  2019-06-10  video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f viewed  2   2019-06-10
   edX/DemoX/Demo_Course_2 honor   2019-06-10  video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f    viewed  2 2019-06-10
   edX/DemoX/Demo_Course   ANONYMOUS USER  2019-06-10  video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f viewed  2   2019-06-10
   edX/DemoX/Demo_Course   honor   2019-06-10  video   i4x-edX-DemoX-video-8c0028eb2a724f48a074bc184cd8635f    viewed  2 2019-06-10
   Time taken: 1.929 seconds, Fetched: 4 row(s)
   ```

**Author notes and concerns**:

1. This change does not modify the various summary tables generated by module_engagement, so the data will only be visible in the hive table itself.
   The summary tables also require that the per-learner data be linked to an enrollment record, which anonymous records cannot be.  So future enhancements would be required to make to make this information visible as appropriate in Insights.
1. Any logs with anonymous user events will need to be re-processed with this option enabled in order to fix the historical data records.

**Reviewers**
- [ ] @gr4yscale


Analytics Pipeline Pull Request
---

Make sure that the following steps are done before merging:

  - [ ] If you have a migration please contact data engineering team before merging.
  - [ ] Before merging run full acceptance tests suite and provide URL for the acceptance tests run.
  - [ ] A member of data engineering team has approved the pull request.
